Refector ElementFinder to better support ctx object.

aaltat opened the pull request:
Also also simplifies where the prefix is determined and which
selenium method is used to find elements

In past //div was defaulted to xpath prefix, because it starts "//".
Now also (// defaults to xpath prefix, because example (//div)[1] is
valid xpath.

aaltat committed:
Moves _parse_locator to ParserLocator class.

Also also simplifies where the prefix is determined and which
selenium method is used to find elements

In past //div was defaulted to xpath prefix, because it starts "//".
Now also (// defaults to xpath prefix, because example (//div)[1] is
valid xpath.

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -4,6 +4,25 @@
 from Selenium2Library.utils import escape_xpath_value, events
 
 
+class ParserLocator(object):
```
`LocatorParser` sounds a much better than `ParserLocator` IMHO. Also, I'd prefer helper class like this to be below the main class of this module.

aaltat replied:
Done

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -4,6 +4,25 @@
+class ParserLocator(object):
+
+    @classmethod
+    def parse_locator(cls, locator):
```
- What benefits does making this a class method bring?
- Since this is a method in a locator parser class, it could be named simply `parse`. Namespaces are a honking good idea and all that.

aaltat replied:
1) The ` @classmethod` allows to use the parser method without creating instance, like `LocatorParser.parse(locator)`
2) Agreed and done

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -4,6 +4,25 @@
+    def parse_locator(cls, locator):
+        prefix = None
+        criteria = locator
+        if locator.startswith('//') or locator.startswith('(//'):
```
`startswith()` accept a tuple of possible matches like `locator.startswith(('//', '(//'))`.

aaltat replied:
I did not know that, good idea. Done

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -23,19 +42,33 @@ def __init__(self):
+            'a': [
+                '@id', '@name', '@href',
+                'normalize-space(descendant-or-self::text())'
+            ],
```
A bit verbose formatting, especially compared to how the earlier entry is on one line only. #nitpick

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -23,19 +42,33 @@ def __init__(self):
-            raise ValueError("Element locator with prefix '" + prefix + "' is not supported")
-        (tag, constraints) = self._get_tag_and_constraints(tag)
+            raise ValueError('Element locator with prefix \'{}\' '
+                             'is not supported'.format(prefix))
```
- Why not double quotes around the string to avoid escaping the internal single quotes?

- As a very minor style thing, instead of using `self._strategies.get()` and validating that result isn't `None`, I'd use

   ```
    try:
        strategy = self._strategies[prefix]
    except KeyError:
        raise ValueError(...)
   ```

   or

   ```
    if prefix not in self._strategies:
        raise ValueError(...)
    strategy = self._strategies[prefix]
    ```

pekkaklarck commented on src/Selenium2Library/locators/elementfinder.py:
```diff
@@ -4,6 +4,25 @@
+                criteria = locator_parts[1].strip()
+            elif len(locator_parts) == 1:
+                prefix = 'default'
+                criteria = locator
```
Something like this would probably be a bit easier to understand:

```python
if locator.startswith(('//', '(//')):
    return 'xpath', locator
if '=' not in locator:
    return 'default', locator
prefix, locator = locator.split('=', 1)
return prefix, locator.strip()
```

pekkaklarck reviewed (approved):
Looks good to me in general. Please look at my line notes, though. Also, was there a possibility that some other piece of the code could use the same locator parsing logic?

aaltat committed:
Fixes based on review

aaltat committed:
Changed to use tuple

aaltat committed:
More fixes based on comments

Commit:
Modified ElementFinder locator parsing and logic.

The _parse_locator now does better determination which strategy
is used from the ElementFinder class. From find method, removed asserts
for empty locator and empty browser, because it does not provide
clear error message. Also contains many formatting changes which
improves code readability in the ElementFinder class.

As last, improved run_tests.py to set pythonpath correctly also in
windows.
